src/cuda/ptx.cu

```diff
@@ -0,0 +1,262 @@
+            break;
+        }
+
+        cudaStatus = cudaMalloc(&d_iv, (to_be_processed_size / 128 ));
+        // iv occupies 32 bytes, piece occupies 4096 bytes.
+        // Instead of dividing the size into 4096, then multiplying it with 32, we can divide into 128.
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "cudaMalloc failed!");
+            break;
+        }
+
+        // computing the next range of pieces to be processed, and copying them into GPU memory
+        cudaStatus = cudaMemcpy(d_piece, (piece + (processed_piece_size / 4)),
+                                to_be_processed_size, cudaMemcpyHostToDevice);
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "cudaMemcpy failed!");
+            break;
+        }
+        cudaStatus = cudaMemcpy(d_iv, (iv + (processed_piece_size / 512)),
+                                (to_be_processed_size / 128), cudaMemcpyHostToDevice);
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "cudaMemcpy failed!");
+            break;
+        }
+        // the reason for the extra division by 4 is:
+        // we are doing pointer arithmetic here. Type of `piece` and `iv` are unsigned int, and unsigned int
+        // is allocating 4 bytes. So actually, iv+1 reaches to next unsigned int, which is 4 bytes later
+        // and we have computed the actual size. We have to divide our computations by 4 in here
+
+        sloth256_189_encode<<<block_count, thread_count>>>(d_piece, d_iv);  // calling the kernel
+
+        cudaStatus = cudaGetLastError();  // Check for any errors launching the kernel
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "Kernel launch failed: %s\n", cudaGetErrorString(cudaStatus));
+            break;
+        }
+
+        cudaStatus = cudaDeviceSynchronize();
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "cudaDeviceSynchronize returned error code %d, aborting...\n", cudaStatus);
+            break;
+        }
+
+        cudaStatus = cudaMemcpy((piece + (processed_piece_size / 4)), d_piece,
+                                to_be_processed_size, cudaMemcpyDeviceToHost);
+        // copy back the result to host, again extra division by 4 because of pointer arithmetic
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "cudaMemcpy failed!");
+            break;
+        }
+
+        processed_piece_size += to_be_processed_size;  // update the processed_piece_size
+        remaining_piece_size -= to_be_processed_size;  // likewise :)
+
+        if (remaining_piece_size == 0) {  // successful!
+            break;  // Hurry! Get out of the loop
+        }
+    }
+
+    cudaFree(d_piece);  // clean-up
+    cudaFree(d_iv);  // clean-up
+
+    return cudaStatus;  // cudaStatus is 0 if there is no error, 1 if there is error
+}
+
+
+extern "C" bool test_1x1_cuda(unsigned int piece[], size_t len,
```

Comment:
Do we need `test` function anymore?

Reply:
I don't think so. I might be beneficial but we can recover it from git history if we need it. Removing...